File: 3852-path-existence-queries-in-a-graph-ii/path-existence-queries-in-a-graph-ii.go
func pathExistenceQueries(n int, nums []int, maxDiff int, queries [][]int) []int {
    idx := make([]int, n)
	pos := make([]int, n)

	for i := 0; i < n; i++ {
		idx[i] = i
	}

	sort.Slice(idx, func(i, j int) bool {
		return nums[idx[i]] < nums[idx[j]]
	})

	for i := 0; i < n; i++ {
		pos[idx[i]] = i
	}

	m := 0
	for t := n; t > 0; t >>= 1 {
		m++
	}

	f := make([][]int, n)
	for i := range f {
		f[i] = make([]int, m)
	}

	left := 0

	for i := 0; i < n; i++ {
		for left < i &&
			nums[idx[i]]-nums[idx[left]] > maxDiff {
			left++
		}

		f[i][0] = left
	}

	for j := 1; j < m; j++ {
		for i := 0; i < n; i++ {
			f[i][j] = f[f[i][j-1]][j-1]
		}
	}

	res := make([]int, 0, len(queries))

	for _, q := range queries {

		x := pos[q[0]]
		y := pos[q[1]]

		if x > y {
			x, y = y, x
		}

		if x == y {
			res = append(res, 0)
			continue
		}

		step := 0

		for i := m - 1; i >= 0; i-- {
			if f[y][i] > x {
				y = f[y][i]
				step += 1 << i
			}
		}

		if f[y][0] <= x {
			res = append(res, step+1)
		} else {
			res = append(res, -1)
		}
	}

	return res
}

File: 3852-path-existence-queries-in-a-graph-ii/path-existence-queries-in-a-graph-ii.py
from typing import List
import math

class Solution:
    def pathExistenceQueries(
        self,
        n: int,
        nums: List[int],
        maxDiff: int,
        queries: List[List[int]]
    ) -> List[int]:

        arr = sorted((v, i) for i, v in enumerate(nums))
        values = [x[0] for x in arr]

        # original index -> sorted position
        pos = [0] * n
        for i, (_, idx) in enumerate(arr):
            pos[idx] = i

        LOG = n.bit_length() + 1

        jump = [[0] * LOG for _ in range(n)]

        r = 0
        for l in range(n):
            while r + 1 < n and values[r + 1] - values[l] <= maxDiff:
                r += 1
            jump[l][0] = r

        for k in range(1, LOG):
            for i in range(n):
                jump[i][k] = jump[jump[i][k - 1]][k - 1]

        def min_jump(start, end):
            if start == end:
                return 0

            if jump[start][0] >= end:
                return 1

            if jump[start][LOG - 1] < end:
                return math.inf

            ans = 0

            for k in range(LOG - 1, -1, -1):
                if jump[start][k] < end:
                    ans += 1 << k
                    start = jump[start][k]

            return ans + 1

        res = []

        for u, v in queries:
            a = pos[u]
            b = pos[v]

            if a > b:
                a, b = b, a

            d = min_jump(a, b)

            if d == math.inf:
                res.append(-1)
            else:
                res.append(d)

        return res